src/training.jl

```diff
@@ -49,18 +49,30 @@ function step! end
 function step!(learner, phase::TrainingPhase, batch)
     xs, ys = batch
     runstep(learner, phase, (; xs=xs, ys=ys)) do handle, state
-        state.grads = gradient(learner.params) do
-            state.ŷs = learner.model(state.xs)
+
+        state.grads, = gradient(learner.model) do model
```

Comment:
I think here you want to take the gradient w.r.t. `learner.params` when the optimizer is a `Flux.Optimise.AbstractOptimiser`. Conversely, if it is not, you take the gradient w.r.t. `learner.model` like you are now.

This is why `update!` below is erroring cause you need to `Grads` object for the old optimizers. And you can only get that with implicit params.

I think some dispatch for the gradient would be easiest. Another option is to have a utility that takes the model, the gradient w.r.t. it, and `Params`, then it produces a `Grads` to match.

Reply:
Ah, I figured. Was hoping there may be a way to have the same `Zygote.gradient` call work but I guess not. I'll add a dispatch on the optimiser there.

Reply:
Well what you're asking for might come eventually in a later version of Flux as part of the AD-agnostic push. So, the code might eventually get simpler.

Reply:
That's good to know. Definitely let me know then, so I can clean this up again.